shared/rate_limits/__init__.py

```diff
@@ -1,17 +1,52 @@
 import logging
+from typing import Optional
 
 from redis import Redis, RedisError
 
-from shared.bots import get_adapter_auth_information
-from shared.bots.types import AdapterAuthInformation
 from shared.config import get_config
 from shared.django_apps.codecov_auth.models import Owner
 from shared.django_apps.core.models import Repository
+from shared.typings.torngit import TorngitInstanceData
 
 log = logging.getLogger(__name__)
 
 
-# Ensure service = "gh" before using
+def gh_app_key_name(installation_id: int, app_id: Optional[int] = None) -> str:
+    app_id = app_id or "default_app"
+    return f"{app_id}_{installation_id}"
+
+
+def owner_key_name(owner_id: int) -> str:
+    return str(owner_id)
+
+
+def bot_key_name() -> str:
+    return str(get_config("github", "bot", "key"))
+
+
+# TODO: Test this
+def determine_entity_redis_key_from_torngit_data(data: TorngitInstanceData) -> str:
+    """
+    Determines entity redis key from torngit data. This is an alternative
+    function to determine_entity_redis_key if the auth information has been
+    parsed into a torngit instance data object.
+
+    It should only be used for github git instances
+    """
+    current_installation = data.get("installation")
+    current_owner = data.get("owner")
+
+    if current_installation and current_installation.get("installation_id"):
+        return gh_app_key_name(
+            app_id=current_installation.get("app_id"),
+            installation_id=current_installation.get("installation_id"),
+        )
+    elif current_owner and current_owner.get("ownerid"):
+        return owner_key_name(owner_id=current_owner.get("ownerid"))
+    else:
+        return bot_key_name()
```

Comment:
I think there's an edge case here when we use one of the TokenTypeMapping tokens in a particular request.

Sometimes we don't use the "main" token in the Torngit instance, but one of the tokens in TokenTypeMapping. These are shared among guest users for specific requests. If _that_ token is rate limited it doesn't mean necessarily that the github_bot is rate limited

This edge case also affects `determine_entity_redis_key` :E

Reply:
Can you help me understand this a bit further? Not sure I fully follow. 

For this and [this](https://github.com/codecov/shared/pull/303#discussion_r1697865195) comment, the goal is to say "this entity (if available) will be rate limited because it's associated token is rate limited". The get_auth_adapter fn in theory does that right, get you who the entity (app/owner) + its token is - I suppose the token itself here doesn't matter as much if it's _always_ going to be the same for the person, but if it rotates + changes, then we'd probably need to adjust the fn to contemplate a combination of app/owner + token